### PR TITLE
Make sure GCC 12 is installed in GCC 12 CI configuration

### PR DESCRIPTION
**Pull request by msimberg**
Testing... but it looks like the "GCC 12" CI configuration was actually using GCC 11. Opening this for debugging and potential fixes. @rasolca and others, please feel free to comment if you have thoughts on how this should be fixed properly.

### CODE REVIEW

**Comment by msimberg** [comment]
cscs-ci run

File: ci/cpu/gcc12_release_cxx20.yml [hunk 1]
```diff
@@ -4,6 +4,7 @@ include:
 cpu gcc12 cxx20 release deps:
   extends: .build_deps_common
   variables:
+    EXTRA_APTGET: "g++-12"
```
Comment (rasolca) [inline ci/cpu/gcc12_release_cxx20.yml:7]: I would add `gfortran-12` as well

**Review by rasolca — CHANGES_REQUESTED** [review]
@albestro can you comment on the fact that spack doesn't complain that the compiler version we ask for doesn't exist?

File: ci/cpu/gcc12_release_cxx20.yml (hunk 1 above)
Comment (albestro) [reply ci/cpu/gcc12_release_cxx20.yml:7]: I'm going to give a check and I'll report here 😉

**Comment by msimberg** [comment]
cscs-ci run

**Comment by albestro** [comment]
> @albestro can you comment on the fact that spack doesn't complain that the compiler version we ask for doesn't exist?

Apparently, we are just asking for a package preference

https://github.com/eth-cscs/DLA-Future/blob/a7900e800000df65c3ed742eb459003da7cca178/ci/docker/build.Dockerfile#L60

which is not a requirement like `packages:all:require:'%${COMPILER}'`.

The question for which I still don't have an answer is: how does the config scope "merging" work in case we add something in the global config and then we have something also in the environment file? I'm going to have a look.

**Comment by albestro** [comment]
> > @albestro can you comment on the fact that spack doesn't complain that the compiler version we ask for doesn't exist?
> 
> Apparently, we are just asking for a package preference
> 
> https://github.com/eth-cscs/DLA-Future/blob/a7900e800000df65c3ed742eb459003da7cca178/ci/docker/build.Dockerfile#L60
> 
> which is not a requirement like `packages:all:require:'%${COMPILER}'`.
> 
> The question for which I still don't have an answer is: how does the config scope "merging" work in case we add something in the global config and then we have something also in the environment file? I'm going to have a look.

https://github.com/spack/spack/pull/36526 This is something important to know that @haampie pointed out. It is a "workaround", but as far as we keep using arrays instead of string, it should work as expected™.

With arrays, we are going to get a set of all rules available in different scopes. Each entry in any scope is considered as an individual rule. If they are not compatible all together, we will get a concretisation error.

**Comment by msimberg** [comment]
cscs-ci run

**Comment by msimberg** [comment]
cscs-ci run

**Comment by msimberg** [comment]
It looks like the GCC 11 version was wrong as well (we specify 11.2.0 but the available version is 11.4.0). I've updated that in the latest commit.

**Comment by albestro** [comment]
> It looks like the GCC 11 version was wrong as well (we specify 11.2.0 but the available version is 11.4.0). I've updated that in the latest commit.

As discussed offline, we might end up indicating just the major version for the compiler (it should just work™), which is actually what the CI general wants (GCC12, CLANG15, ...).

**Comment by msimberg** [comment]
cscs-ci run

**Review by rasolca — APPROVED** [review]

**Pull request merged**